src/librustc/ty/layout.rs

```diff
@@ -2108,142 +1988,179 @@ impl<'a, 'tcx> SizeSkeleton<'tcx> {
 
-    fn normalize_projections(self, ty: Ty<'tcx>) -> Ty<'tcx> {
-        self.tcx.normalize_associated_type_in_env(&ty, self.param_env)
+        Ok(layout)
     }
 }
 
-impl<'a, 'tcx> TyLayout<'tcx> {
-    pub fn for_variant(&self, variant_index: usize) -> Self {
-        TyLayout {
-            variant_index: Some(variant_index),
-            ..*self
-        }
-    }
+impl<'a, 'tcx> LayoutOf<Ty<'tcx>> for (ty::maps::TyCtxtAt<'a, 'tcx, 'tcx>,
+                                       ty::ParamEnv<'tcx>) {
+    type TyLayout = Result<TyLayout<'tcx>, LayoutError<'tcx>>;
 
-    pub fn field_offset<C: HasDataLayout>(&self, cx: C, i: usize) -> Size {
-        self.layout.field_offset(cx, i, self.variant_index)
-    }
+    /// Computes the layout of a type. Note that this implicitly
+    /// executes in "reveal all" mode.
+    #[inline]
+    fn layout_of(self, ty: Ty<'tcx>) -> Self::TyLayout {
+        let (tcx_at, param_env) = self;
 
-    pub fn field_count(&self) -> usize {
-        // Handle enum/union through the type rather than Layout.
-        if let ty::TyAdt(def, _) = self.ty.sty {
-            let v = self.variant_index.unwrap_or(0);
-            if def.variants.is_empty() {
-                assert_eq!(v, 0);
-                return 0;
-            } else {
-                return def.variants[v].fields.len();
-            }
-        }
+        let ty = tcx_at.tcx.normalize_associated_type_in_env(&ty, param_env.reveal_all());
+        let cached = tcx_at.layout_raw(param_env.reveal_all().and(ty))?;
+        let layout = TyLayout {
+            ty,
+            cached
+        };
 
-        match *self.layout {
-            Scalar { .. } => {
-                bug!("TyLayout::field_count({:?}): not applicable", self)
-            }
+        // NB: This recording is normally disabled; when enabled, it
+        // can however trigger recursive invocations of `layout_of`.
+        // Therefore, we execute it *after* the main query has
+        // completed, to avoid problems around recursive structures
+        // and the like. (Admitedly, I wasn't able to reproduce a problem
+        // here, but it seems like the right thing to do. -nmatsakis)
+        CachedLayout::record_layout_for_printing(tcx_at.tcx, ty, param_env, layout);
 
-            // Handled above (the TyAdt case).
-            CEnum { .. } |
-            General { .. } |
-            UntaggedUnion { .. } |
-            RawNullablePointer { .. } |
-            StructWrappedNullablePointer { .. } => bug!(),
+        Ok(layout)
+    }
+}
 
-            FatPointer { .. } => 2,
+impl<'a, 'tcx> TyLayout<'tcx> {
+    pub fn for_variant<C>(&self, cx: C, variant_index: usize) -> Self
+        where C: LayoutOf<Ty<'tcx>> + HasTyCtxt<'tcx>,
+              C::TyLayout: MaybeResult<TyLayout<'tcx>>
+    {
+        let cached = match self.variants {
+            Variants::Single { index } if index == variant_index => self.cached,
+
+            Variants::Single { index } => {
```

Comment:
this looks like it's going to leak a layout for each time you call `for_variant` rather than being cached.

Reply:
`intern_layout` is used so the allocation only happens once, cached by variant index and number of fields. Also, this *only* happens for accessing uninhabited variants.